docs/RFCS/20220524_sso_authentication.md

```diff
@@ -0,0 +1,145 @@
+* Service accounts, hosted in CC or in the customer's identity provider, are 
+able to create a CRDB session. 
+
+The following are out of scope:
+* Authorization e.g. user permissions propagation from CC to CRDB
+* Just-in-time provisioning of SQL users at login time
+* SSO for DB Console
+
+# Motivation
+
+Customers are currently unable to enforce their enterprise authentication policies and 
+are unable to ensure a single point of access across the product for their users.
+
+As support for CC SSO grows, it is crucial that we provide SSO for CRDB as well 
+to prevent any confusion for organization admins regarding seeing
+different authentication methods across the product. 
+
+# Background
+
+Currently, clients can authenticate themselves to the server in the following ways:
+* Password 
+* SCRAM 
+  * A password hashing protocol that offers multiple security benefits 
+  over password.
+  * Recommended over password, assuming clients can support it (many do not).
+* Certificate
+  * A certificate/key pair that is signed by a certificate authority is used to 
+  authenticate the client.
+  * Not supported for Cloud instances.
+* Trust (and Reject)
+  * Always allows access or never allow access.
+  * Usually paired with restrictions such as requiring requests come 
+  from specific IP addresses or unix sockets
+* Kerberos / GSSAPI
+    * Not supported for Cloud instances.
+
+This RFC proposes to add another authentication method that will have access 
+tokens, more specifically JSON Web Tokens (JWTs), to be passed into the password field, 
+and this new method will achieve the aforementioned goals.
+
+# Technical Design
+
+The token credentials will consist of an access token and a refresh token. The access 
+tokens will be signed and short-lived while the refresh tokens will be opaque and 
+long-lived. 
+
+Refresh tokens can be sent to the credential issuer to retrieve a new set of 
+token credentials. 
+
+The CRDB CLI login flow using tokens as credentials will be this:
+1. CRDB CLI redirects the user to CC Console to complete the login flow.
+2. A set of access and refresh token credentials are generated.
+3. The access token is passed as an authentication credential into the password field in CLI.
+4. CRDB recognizes that the credential is token-based and validates the 
+user, based on which either the request succeeds or fails.
+
+For application clients, one difference is that the user directly fetches the token credentials 
+issued by CC instead of receiving it. 
+
+**Add a SQL authentication method that uses access tokens** 
+
+A new login method will be created that will allow an access token to be 
+presented as a password during the normal login flow. Recognizing that
+it is a token, CRDB will then verify the token's signature and will accept 
+it as authentication. This verification will happen without a network call.
+
+**Add a CC API endpoint to refresh login tokens**
```

Comment:
I think this needs a bit more detail on how the endpoint is implemented. Who will generate the token and how? Where are the tokens stored? What is the format of the token? How are the tokens verified?

Reply:
The state of the PR was an early WIP at that point, more details now.